### PR TITLE
chore: add TAV tests for mongoose

### PR DESCRIPTION
**Pull request by david-luna**
# Description

Adding a new test for `mongoose` package and update `tav.yml` file accordingly. With this new test were going to check the switch from `mongodb-core` to `mongodb` as dependency doesn't affect the instrumentation and also we will have checks on the versions we claim to support in the docs.

Refs: https://github.com/elastic/apm-agent-nodejs/issues/3161

### Checklist

- [ ] Implement code
- [x] Add tests
- [ ] Update TypeScript typings
- [ ] Update documentation
- [ ] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by david-luna** [comment]
It seems TAV tests did not run after my command. Running it locally it errors with node v18 and mongoose v6.11.1. Logs attached

```
-- installing ["mongoose@6.11.1"]
-- running test "node test/instrumentation/modules/mongoose.test.js" with mongoose
(node:44099) [MONGOOSE] DeprecationWarning: Mongoose: the `strictQuery` option will be switched back to `false` by default in Mongoose 7. Use `mongoose.set('strictQuery', false);` if you want to prepare for this change. Or use `mongoose.set('strictQuery', true);` to suppress this warning.
(Use `node --trace-deprecation ...` to show where the warning was created)
/Users/david/Documents/repos/el/apm-agent-nodejs/test/instrumentation/modules/mongoose.test.js:65
    t.equal(data.transactions[0].name, 't0', 'transaction.name')
                                 ^

TypeError: Cannot read properties of undefined (reading 'name')
    at /Users/david/Documents/repos/el/apm-agent-nodejs/test/instrumentation/modules/mongoose.test.js:65:34
    at Object._write (/Users/david/Documents/repos/el/apm-agent-nodejs/test/_mock_http_client.js:46:50)
    at Object.sendSpan (/Users/david/Documents/repos/el/apm-agent-nodejs/test/_mock_http_client.js:50:12)
    at /Users/david/Documents/repos/el/apm-agent-nodejs/lib/instrumentation/index.js:481:28
    at done (/Users/david/Documents/repos/el/apm-agent-nodejs/lib/instrumentation/span.js:486:5)
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)

Node.js v18.16.0
-- detected failing command, flushing stdout...
{"log.level":"info","@timestamp":"2023-06-07T14:58:28.710Z","log":{"logger":"elastic-apm-node"},"agentVersion":"3.46.0","env":{"pid":44099,"proctitle":"node","os":"darwin 22.5.0","arch":"arm64","host":"Davids-MBP.home","timezone":"UTC+0200","runtime":"Node.js v18.16.0"},"config":{"serviceName":{"source":"start","value":"test-mongoose","commonName":"service_name"},"serviceVersion":{"source":"default","value":"3.46.0","commonName":"service_version"},"serverUrl":{"source":"default","value":"http://127.0.0.1:8200/","commonName":"server_url"},"logLevel":{"source":"default","value":"info","commonName":"log_level"},"captureExceptions":{"source":"start","value":false},"metricsInterval":{"source":"start","value":0},"centralConfig":{"source":"start","value":false},"cloudProvider":{"source":"start","value":"none"},"spanCompressionEnabled":{"source":"start","value":false}},"activationMethod":"require","ecs":{"version":"1.6.0"},"message":"Elastic APM Node.js Agent v3.46.0"}
TAP version 13
# Mongoose simple test
{
  length: 5,
  spans: [
    {
      id: 'aad5c2c75e9724ba',
      transaction_id: '47ecfc0a87e486a1',
      parent_id: '47ecfc0a87e486a1',
      trace_id: 'f837ca3124e7028ebe3e5db769514b4a',
      name: 'elasticapm.tests.create',
      type: 'db',
      subtype: 'mongodb',
      action: 'create',
      timestamp: 1686149908941532,
      duration: 1,
      context: [Object],
      stacktrace: undefined,
      sync: true,
      outcome: 'success',
      sample_rate: 1
    },
    {
      id: '693fc795e32a51b3',
      transaction_id: '47ecfc0a87e486a1',
      parent_id: '47ecfc0a87e486a1',
      trace_id: 'f837ca3124e7028ebe3e5db769514b4a',
      name: 'elasticapm.tests.insert',
      type: 'db',
      subtype: 'mongodb',
      action: 'insert',
      timestamp: 1686149908942156,
      duration: 2,
      context: [Object],
      stacktrace: undefined,
      sync: true,
      outcome: 'success',
      sample_rate: 1
    },
    {
      id: '6e9fdb846c175d2f',
      transaction_id: '47ecfc0a87e486a1',
      parent_id: '47ecfc0a87e486a1',
      trace_id: 'f837ca3124e7028ebe3e5db769514b4a',
      name: 'elasticapm.tests.find',
      type: 'db',
      subtype: 'mongodb',
      action: 'find',
      timestamp: 1686149908946070,
      duration: 1,
      context: [Object],
      stacktrace: undefined,
      sync: false,
      outcome: 'success',
      sample_rate: 1
    },
    {
      id: '3d627a4699414f3b',
      transaction_id: '47ecfc0a87e486a1',
      parent_id: '47ecfc0a87e486a1',
      trace_id: 'f837ca3124e7028ebe3e5db769514b4a',
      name: 'elasticapm.tests.delete',
      type: 'db',
      subtype: 'mongodb',
      action: 'delete',
      timestamp: 1686149908948012,
      duration: 0,
      context: [Object],
      stacktrace: undefined,
      sync: false,
      outcome: 'success',
      sample_rate: 1
    },
    {
      id: '23ba0c8a6089f5c2',
      transaction_id: '47ecfc0a87e486a1',
      parent_id: '47ecfc0a87e486a1',
      trace_id: 'f837ca3124e7028ebe3e5db769514b4a',
      name: 'admin.$cmd.endSessions',
      type: 'db',
      subtype: 'mongodb',
      action: 'endSessions',
      timestamp: 1686149908949448,
      duration: 0,
      context: [Object],
      stacktrace: undefined,
      sync: true,
      outcome: 'success',
      sample_rate: 1
    }
  ],
  transactions: [],
  errors: [],
  metricsets: []
}

-- fatal: Test exited with code 1
```

File: .tav.yml
Comment (trentm) [inline .tav.yml:177]: Probably want a `... <8` specifier on here so we test new 7.x releases.  Probably for the 5.x and 6.x series as well.

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:59]: Debugging code to remove eventually?

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:61]: Is the 4th one a span for the `.deleteMany()` that we don't want to test?

**Review by trentm — COMMENTED** [review]
This looks good, but you say that TAV tests are still failing? I assume you want to dig into those before merging, right?

**Comment by trentm** [comment]
> It seems TAV tests did not run after my command.

The difference with mine at https://github.com/elastic/apm-agent-nodejs/pull/3408#pullrequestreview-1468392169 is the `/test tav ...` comment needs to be on a *review comment*, because it triggers on `pull_request_review`: https://github.com/elastic/apm-agent-nodejs/blob/a3b9a60c6fc6d4a9a8cffc57b27997abd3680092/.github/workflows/tav-command.yml#L4

(Also we'd had a case where tav-command did not run this past week because of permissions issues. However, it seems to be working this time, so I'm not sure what is/was going on with permissions. /cc @cachedout  My tav-command attempt worked on this PR this time.)

**Review by trentm — COMMENTED** [review]
/test tav mongoose

**Comment by trentm** [comment]
> Running it locally it errors with node v18 and mongoose v6.11.1.

Likewise for node v16 and mongoose v6.11.1.

File: test/instrumentation/modules/mongoose.test.js
Comment (david-luna) [reply test/instrumentation/modules/mongoose.test.js]: Yes!

File: test/instrumentation/modules/mongoose.test.js
Comment (david-luna) [reply test/instrumentation/modules/mongoose.test.js]: the last span seems to be an internal operation triggered by `mongoose` after it executes all we requested (insert, find & delete). Name of the span is `admin.$cmd.command`. You already commented later.

File: .ci/tav.json
Comment (trentm) [inline .ci/tav.json:26]: THis puts us at 34 * 7 = 238 combinations, so still under the 256 GH actions jobs limit. All good.

File: package.json
Comment (trentm) [inline package.json:110]: Oh, oops, this is in "dependencies" instead of "devDependencies". That needs to change.

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:76]: Actually, I believe you can just `await agent.flush()`.

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:94]: I'm fine with having the test just assert that all the spans are there and not requiring the order to be as expected... with a comment here saying why.

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:103]: Do we know the versions for which we expect the `create` span?

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:96]: Perhaps it would be clearer to explicitly mention and filter out the `admin.*` spans.

**Review by trentm — CHANGES_REQUESTED** [review]

**Comment by trentm** [comment]
It looks like all the tests with `contextManager=patch` are failing. That is not too surprising. We can just skip tests in that case (https://github.com/elastic/apm-agent-nodejs/blob/main/test/instrumentation/modules/@aws-sdk/client-s3.test.js#L14-L17 is an example of doing this) and perhaps add a short mention in supported-technologies.asciidoc that mongoose isn't supported with the deprecated `contextManager=patch` setting.

**Review by trentm — COMMENTED** [review]
/test tav mongoose

File: package.json
Comment (david-luna) [reply package.json:110]: My bad. maybe in one of the tests I've used ' --save` instead of `--save-dev`

Comment (david-luna) [reply package.json]: Fixed!

File: test/instrumentation/modules/mongoose.test.js
Comment (david-luna) [reply test/instrumentation/modules/mongoose.test.js:94]: I thought so since the order is internal to `mongoose` package but it's alway nice to have a second opinion :)

File: test/instrumentation/modules/mongoose.test.js
Comment (david-luna) [reply test/instrumentation/modules/mongoose.test.js:103]: I did some checks and seems that `v6.0.6` introduced that command

File: dev-utils/gen-notice.sh
Comment (trentm) [inline dev-utils/gen-notice.sh:1]: Now that mongoose is in devDeps, I think the change in this file can be dropped as well.

**Comment by david-luna** [comment]
> It looks like all the tests with `contextManager=patch` are failing. That is not too surprising. We can just skip tests in that case (https://github.com/elastic/apm-agent-nodejs/blob/main/test/instrumentation/modules/@aws-sdk/client-s3.test.js#L14-L17 is an example of doing this) and perhaps add a short mention in supported-technologies.asciidoc that mongoose isn't supported with the deprecated `contextManager=patch` setting.

Thanks for the example. Although were not instrumenting mongoose directly do you think it's possible show a warning in this situation?

**Comment by david-luna** [comment]
TAV tests reveal there is an error instrumenting mongoose on a certain combination of versions
- mongoose <5.7.0
- nodejs <=10

Here's the test output
```shell
> elastic-apm-node@3.46.0 test:tav /Users/david/Documents/repos/el/apm-agent-nodejs
> (cd test/opentelemetry-metrics/fixtures && tav --quiet) && (cd test/opentelemetry-bridge && tav --quiet) && (cd test/instrumentation/modules/next/a-nextjs-app && tav --quiet) && tav --quiet

-- ok
-- ok
-- ok
-- required packages ["mongoose@5.7.0"]
-- installing ["mongoose@5.7.0"]
-- running test "node test/instrumentation/modules/mongoose.test.js" with mongoose
-- required packages ["mongoose@5.6.13"]
-- installing ["mongoose@5.6.13"]
-- running test "node test/instrumentation/modules/mongoose.test.js" with mongoose
(node:82390) DeprecationWarning: current URL string parser is deprecated, and will be removed in a future version. To use the new parser, pass option { useNewUrlParser: true } to MongoClient.connect.
-- detected failing command, flushing stdout...
{"log.level":"info","@timestamp":"2023-06-09T18:30:17.582Z","log":{"logger":"elastic-apm-node"},"agentVersion":"3.46.0","env":{"pid":82390,"proctitle":"node","os":"darwin 22.5.0","arch":"x64","host":"Davids-MBP.home","timezone":"UTC+0200","runtime":"Node.js v10.24.1"},"config":{"serviceName":{"source":"start","value":"test-mongoose","commonName":"service_name"},"serviceVersion":{"source":"default","value":"3.46.0","commonName":"service_version"},"serverUrl":{"source":"default","value":"http://127.0.0.1:8200/","commonName":"server_url"},"logLevel":{"source":"default","value":"info","commonName":"log_level"},"captureExceptions":{"source":"start","value":false},"metricsInterval":{"source":"start","value":0},"centralConfig":{"source":"start","value":false},"cloudProvider":{"source":"start","value":"none"},"spanCompressionEnabled":{"source":"start","value":false}},"activationMethod":"require","ecs":{"version":"1.6.0"},"message":"Elastic APM Node.js Agent v3.46.0"}
TAP version 13
# Mongoose simple test
ok 1 no error from agent.flush()
ok 2 transaction.name
ok 3 spans parent_id
ok 4 spans subtype
ok 5 insert span present
ok 6 find span present
not ok 7 delete span present
  ---
    operator: ok
    expected: true
    actual:   false
    at: <anonymous> (/Users/david/Documents/repos/el/apm-agent-nodejs/test/instrumentation/modules/mongoose.test.js:97:7)
    stack: |-
      Error: delete span present
          at Test.assert [as _assert] (/Users/david/Documents/repos/el/apm-agent-nodejs/node_modules/tape/lib/test.js:312:48)
          at Test.bound [as _assert] (/Users/david/Documents/repos/el/apm-agent-nodejs/node_modules/tape/lib/test.js:95:17)
          at Test.assert (/Users/david/Documents/repos/el/apm-agent-nodejs/node_modules/tape/lib/test.js:431:7)
          at Test.bound [as ok] (/Users/david/Documents/repos/el/apm-agent-nodejs/node_modules/tape/lib/test.js:95:17)
          at /Users/david/Documents/repos/el/apm-agent-nodejs/test/instrumentation/modules/mongoose.test.js:97:7
          at process._tickCallback (internal/process/next_tick.js:68:7)
  ...

1..7
# tests 7
# pass  6
# fail  1
```

A bit of debugging reveals the `remove` query is intercepted by the agent but there is no parent transaction. Latest in `mongoose@^5.6.0` is 5.6.13 released in 2019-09-06. We should consider to not give support to such old versions

**Comment by trentm** [comment]
> A bit of debugging reveals the `remove` query is intercepted by the agent but there is no parent transaction. Latest in `mongoose@^5.6.0`

This does not explain https://github.com/elastic/apm-agent-nodejs/issues/3161 then right? That user said they are using `"mongoose": "^6.9.1"`

> We should consider to not give support to such old versions

I agree that we do not need to spend time debugging such an old version.

**Comment by trentm** [comment]
> Although were not instrumenting mongoose directly do you think it's possible show a warning in this situation?

I think that would require either instrumenting mongoose (too much added code just for this warning) or putting the warning in the mongodb instrumentation (if that is appropriate).

File: test/instrumentation/modules/mongoose.test.js
Comment (trentm) [inline test/instrumentation/modules/mongoose.test.js:80]: This could be:

```js
await agent.flush()
// ... then dedent the following code that was in the `.then(...)` callback.
```

**Comment by david-luna** [comment]
@trentm these tests were made to catch the problem defined in #3161 and we already discovered is an issue in `mongodb. I'm inclined to close this PR. WDYT?

**Comment by trentm** [comment]
> I'm inclined to close this PR. WDYT?

Fine with me.

It would be nice to have a separate commit that:
- updates the supported mongoose versions in "supported-technologies.asciidoc", and
- perhaps keep the examples/trace-mongoose.js that was written

**Comment by david-luna** [comment]
> It would be nice to have a separate commit that:
>
> - updates the supported mongoose versions in "supported-technologies.asciidoc", and
> - perhaps keep the examples/trace-mongoose.js that was written

Commit in https://github.com/elastic/apm-agent-nodejs/pull/3653

**Pull request closed without merging**
